clevertap-core/src/main/java/com/clevertap/android/sdk/usereventlogs/UserEventLogDAOImpl.kt

```diff
@@ -0,0 +1,375 @@
+        }
+    }
+
+    @WorkerThread
+    override fun cleanUpExtraEvents(threshold: Int, numberOfRowsToCleanup: Int): Boolean {
+        if (threshold <= 0) {
+            logger.verbose("Invalid threshold value: $threshold. Threshold should be greater than 0")
+            return false
+        }
+        if (numberOfRowsToCleanup < 0) {
+            logger.verbose("Invalid numberOfRowsToCleanup value: $numberOfRowsToCleanup. Should be greater than or equal to 0")
+            return false
+        }
+        if (numberOfRowsToCleanup >= threshold) {
+            logger.verbose("Invalid numberOfRowsToCleanup value: $numberOfRowsToCleanup. Should be less than threshold: $threshold")
+            return false
+        }
+
+        val tName = table.tableName
+        val numberOfRowsToKeep = threshold - numberOfRowsToCleanup
+
+        return try {
+            // SQL query to delete only the least recently used rows, using a subquery with LIMIT
+            // When above threshold is reached, delete in such a way that (threshold - numberOfRowsToCleanup) rows exists after cleanup
+            val query = """
+            DELETE FROM $tName
+            WHERE (${Column.NORMALIZED_EVENT_NAME}, ${Column.DEVICE_ID}) IN (
+                SELECT ${Column.NORMALIZED_EVENT_NAME}, ${Column.DEVICE_ID}
+                FROM $tName
+                ORDER BY ${Column.LAST_TS} ASC 
+                LIMIT (
+                SELECT CASE 
+                    WHEN COUNT(*) > ? THEN COUNT(*) - ?
+                    ELSE 0
+                END 
+                FROM $tName
+                )
+            );
+        """.trimIndent()
+
+            // Execute the delete query with the threshold as an argument
+            db.writableDatabase.execSQL(query, arrayOf(threshold,numberOfRowsToKeep))
+            logger.verbose("If row count is above $threshold then only keep $numberOfRowsToKeep rows in $tName")
+            true
+        } catch (e: Exception) {
+            logger.verbose("Error cleaning up extra events in $tName.", e)
+            false
+        }
+    }
+
+    @WorkerThread
+    private fun <T> readEventColumnByDeviceIdAndNormalizedEventName(
+        deviceID: String,
+        normalizedEventName: String,
+        column: String,
+        defaultValueExtractor: () -> T,
+        valueExtractor: (cursor: Cursor, columnName: String) -> T
+    ): T {
+        val tName = table.tableName
+        val selection = "${Column.DEVICE_ID} = ? AND ${Column.NORMALIZED_EVENT_NAME} = ?"
+        val selectionArgs = arrayOf(deviceID, normalizedEventName)
+        val projection = arrayOf(column)
+
+        return try {
+            db.readableDatabase.query(
+                tName, projection, selection, selectionArgs, null, null, null, null
+            )?.use { cursor ->
+                if (cursor.moveToFirst()) {
+                    valueExtractor(cursor, column)
+                } else {
+                    defaultValueExtractor()
```

Comment:
I believe we need to return 0 if there are no records.

Reply:
resolved by @vasct